### PR TITLE
Delayed desktop start

### PR DESCRIPTION
**Pull request by zardus**
Currently, we start the desktop on container start, but many people use vscode or ssh instead. Rather than have hundreds of vnc/X/XFCE processes idling at all times, we should hold off on starting them until they're necessary. This PR does that.

Pro:
- less overall resource usage
- faster container starts!

Con:
- the first visit to `/workspace/desktop` after a container start takes longer (because it waits until `Xtigervnc` and the websocket stuff is started)
- we had to ditch fluxbox because it started crashing on xterm launch in this configuration. This is probably for the best, as XFCE is what we use in production, so that's what we should run testcases with. Let's see how it affects test times...

### CODE REVIEW

**Comment by zardus** [comment]
Looks like this increases CI times by 30 seconds to a minute... Probably worth it to be closer to production?

**Comment by zardus** [comment]
With this PR merged, this is what a fresh container will look like:

```
yans@ramoth ~ $ docker exec -it dojo-test docker exec -it --user=root user_1 ps aux
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
hacker         1  0.0  0.0   1136     4 ?        Ss   06:56   0:00 /sbin/docker-init -- /bin/sleep 6h
hacker         7  0.0  0.0   2744   576 ?        S    06:56   0:00 /bin/sleep 6h
root          98  0.0  0.0   6128  3028 pts/0    Rs+  06:56   0:00 ps aux
yans@ramoth ~ $ 
```

**Comment by zardus** [comment]
@robwaz would you mind looking over the windows-specific changes and maybe testing them?

**Review by ConnorNelson — CHANGES_REQUESTED** [review]
Overall this sounds great for minimizing resource usage. 

I think we can make this a lot smoother by auto starting services that have been recently used. If a service touches a file in .dojo within the users home directory, and uses the access time to decide on start if it should auto start it. We can auto start services used in the last 10 days or whatever. 

This can of course be a future change, though it might address your testcase concerns? Not sure how often we create a new user for a testcase.

File: dojo_plugin/pages/workspace.py [hunk 1]
```diff
@@ -39,6 +40,12 @@ def view_desktop():
     if not container:
         return render_template("iframe.html", active=False)
 
+    exec_run(
+        "/opt/pwn.college/services.d/desktop 2>&1 > /tmp/.dojo/service-desktop.log",
+        user="hacker", pwncollege_uid=user.id, shell=True,
```
[the remark this answers is not in the document]
Comment (zardus) [reply dojo_plugin/pages/workspace.py:45]: You're not concerned about the confusion between `user` (docker user) and `user_id` (pwncollege user)?

File: challenge/services.d/desktop-windows [hunk 2]
```diff
@@ -14,3 +16,5 @@ start-stop-daemon --start \
                   </dev/null \
                   >>/tmp/.dojo/vnc/websockify-windows.log \
                   2>&1
+
+until curl -s dojo-user:6082 >/dev/null; do sleep 0.1; done
```
[the remark this answers is not in the document]
Comment (zardus) [reply challenge/services.d/desktop-windows:21]: For the desktop, we wait for X to start up, which takes a bit longer than everything else anyways. I added the check anyways for consistency, though.

**Comment by zardus** [comment]
Regarding the testcase runtime concerns, I think the test times for this guy are within the range of older test times. The big thing that impacts test times is the merging of `builder-desktop-base` and `builder-desktop-xfce` because fluxbox is buggy; it's just more stuff to install, but the installs are pretty fast as the apt cache seems to be local in the runner.

Definitely agree that we can be smarter about what to start. We don't even need to do crazy shenanigans with the home directory: we know when users click on "Workspace", "Desktop", and so on; we can even just (ab)use the redis cache to store a week-long cookie regarding what services to start for the user, purely on the dojo_plugin side of things.

File: dojo_plugin/pages/workspace.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zardus) [reply dojo_plugin/pages/workspace.py:45]: We can also do `user_id` and `docker_user`.

**Comment by ConnorNelson** [comment]
Ah that makes more sense for the testcase runtime.

Yeah that could also work. Not clear to me which design makes more sense; if CTFd should hold that preemptive responsibility, or if the workspace container itself should. Of course both work.

As for "docker_user"; probably "workspace_user" makes even more sense. I'm increasingly thinking as an abstraction, the user container runtime (and all associated services/ways of accessing it) is "the workspace". Of course right now we currently call VSCode "Workspace", but I think this is a bad terminology; VSCode is just one way of accessing the workspace, as is the desktop, as is ssh. Not immediately relevant to this PR, but I think this is something we should better convey from a UX perspective; "when you start a challenge, the dojo prepares your workspace for that challenge".

**Comment by zardus** [comment]
I'm not sure who should hold it, but I think the implementation will be less error-prone on the ctfd side. I'm down with whatever (when we do it).

I agree with the terminology... It'd be interesting to explore other UX options. Here's a 10-second mock-up for one potential option: 
image

Anyways, I updated `docker_user` to `workspace_user`. We ready to merge?

**Comment by ConnorNelson** [comment]
Yeah let's merge.

I don't think small buttons for critical UX paths make sense. I was thinking about possibly going back to the very original idea of desktop inside vscode. It would be a simple extension that just iframes out to the desktop page. I think that could be super slick (and allows a nice back and forth between the two). We'd need to see how that feels. Alternatively some sort of tabbing or drop down menu or something. We have to also think that we probably want to make `windows-desktop` fit in here, and I've been thinking about the "challenge" service, for hypothetical future challenges that wan't to present their own custom web interface that forwards through (e.g. the phone phreaking challenge). VSCode/Desktop are always available, conditionally more web interfaces are as well. Either way, separate concern.

**Comment by zardus** [comment]
One note after this change: the dojo runs the service init script on every connection, which is fine for novnc and vscode (they both use long-running connections), but it's something to keep in mind if we have other services that result in lots of connections.

**Pull request merged**